File: app/src/main/java/com/quigglesproductions/secureimageviewer/ui/enhancedfileviewer/EnhancedFileViewFragment.java
package com.quigglesproductions.secureimageviewer.ui.enhancedfileviewer;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowInsets;
import android.view.WindowInsetsController;
import android.view.WindowManager;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.OptIn;
import androidx.core.view.ActionProvider;
import androidx.core.view.WindowCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.media3.common.util.UnstableApi;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.quigglesproductions.secureimageviewer.R;
import com.quigglesproductions.secureimageviewer.managers.FolderManager;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.EnhancedFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.EnhancedFolder;
import com.quigglesproductions.secureimageviewer.ui.EnhancedMainMenuActivity;
import com.quigglesproductions.secureimageviewer.ui.IFileViewer;
import com.quigglesproductions.secureimageviewer.ui.SecureActivity;
import com.quigglesproductions.secureimageviewer.ui.SupportActionBarSetListener;
import com.quigglesproductions.secureimageviewer.ui.compoundcontrols.FileViewerNavigator;

import java.lang.reflect.Field;
@OptIn(markerClass = UnstableApi.class)
public class EnhancedFileViewFragment extends Fragment implements IFileViewer {
    private int startPos;
    ViewPager2 viewPager;
    LinearLayout topLayout;
    ImageButton backButton;
    TextView fileName;
    FileViewerNavigator fileNavigator;
    EnhancedFile selectedFile;
    int currentPagerSlopMultiplier;
    boolean hasStartPosition;
    EnhancedFileViewerViewModel viewModel;
    public EnhancedFileViewFragment(){
        hasStartPosition = false;
    }
    public EnhancedFileViewFragment(int startPosition){
        startPos = startPosition;
        hasStartPosition = true;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        viewModel = new ViewModelProvider(this).get(EnhancedFileViewerViewModel.class);
        if(((EnhancedMainMenuActivity)requireActivity()).getSupportActionBar() != null) {
            ((EnhancedMainMenuActivity) requireActivity()).getSupportActionBar().hide();
        }
        else{
            ((EnhancedMainMenuActivity) requireActivity()).registerActionBarSetListener(new SupportActionBarSetListener() {
                @Override
                public void SupportActionBarSet() {
                    ((EnhancedMainMenuActivity) requireActivity()).getSupportActionBar().hide();
                }
            });
        }
        ContextThemeWrapper wrapper = new ContextThemeWrapper(requireContext(),R.style.FileViewerTheme);
        LayoutInflater themedInflater = inflater.cloneInContext(wrapper);
        ((EnhancedMainMenuActivity)getActivity()).hideStatusBar();
        return themedInflater.inflate(R.layout.fragment_file_pager, container, false);
    }

    @Override
    public void onResume() {
        super.onResume();

    }

    @Override
    public void onPause() {
        super.onPause();

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        viewPager = view.findViewById(R.id.fragment_view_pager);
        if(!hasStartPosition)
            startPos = EnhancedFileViewFragmentArgs.fromBundle(getArguments()).getStartPosition();
        setupNavigationControls(view);
        EnhancedFileCollectionAdapter collectionAdapter = new EnhancedFileCollectionAdapter(this);
        collectionAdapter.setFileNavigator(fileNavigator);
        EnhancedFolder selectedFolder = FolderManager.getInstance().getCurrentFolder();
        collectionAdapter.addFiles(selectedFolder.getBaseItems());
        viewPager.setAdapter(collectionAdapter);
        super.onViewCreated(view, savedInstanceState);
        hideSystemBars();
        fileNavigator.setFileTotal(collectionAdapter.getItemCount());
        viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageScrollStateChanged(int state) {
                super.onPageScrollStateChanged(state);
            }

            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);
                selectedFile = collectionAdapter.getItem(position);
                fileName.setText(selectedFile.getName());
                fileNavigator.setFilePosition(position+1);
                topLayout.invalidate();
            }

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                super.onPageScrolled(position, positionOffset, positionOffsetPixels);
            }
        });
        viewPager.setCurrentItem(startPos,false);
        selectedFile = collectionAdapter.getItem(startPos);
        fileName.setText(selectedFile.getName());
        viewPager.setNestedScrollingEnabled(true);
        collectionAdapter.setFileZoomLevelCallback(new EnhancedFileCollectionAdapter.ZoomLevelChangeCallback() {
            @Override
            public void zoomLevelChanged(boolean isZoomed) {
                if(isZoomed)
                    setViewPagerSlop(11);
                else
                    setViewPagerSlop(1);
                return;
            }
        });

    }

    private void setViewPagerSlop(int multiplier){
        int initialValue = 42;

        try {
            if(currentPagerSlopMultiplier != multiplier) {
                Field recyclerViewField = viewPager.getClass().getDeclaredField("mRecyclerView");
                recyclerViewField.setAccessible(true);
                RecyclerView recyclerView = (RecyclerView) recyclerViewField.get(viewPager);
                Field field = RecyclerView.class.getDeclaredField("mTouchSlop");
                field.setAccessible(true);
                field.set(recyclerView, initialValue * multiplier);
            }
        } catch (NoSuchFieldException | IllegalAccessException e) {
            e.printStackTrace();
        }finally {
            currentPagerSlopMultiplier = multiplier;
        }
    }
    private void setupNavigationControls(View view){
        topLayout = view.findViewById(R.id.topLinearLayout);
        backButton = topLayout.findViewById(R.id.backButton);
        fileName = topLayout.findViewById(R.id.file_name);
        fileNavigator = view.findViewById(R.id.fileviewer_navigator);
        backButton.setOnClickListener(new View.OnClickListener() {
            @SuppressLint("RestrictedApi")
            @Override
            public void onClick(View view) {
                getActivity().onBackPressed();
            }
        });
        fileNavigator.setNextButtonOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                viewPager.setCurrentItem(viewPager.getCurrentItem()+1,true);
            }
        });
        fileNavigator.setPreviousButtonOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                viewPager.setCurrentItem(viewPager.getCurrentItem()-1,true);
            }
        });
        fileNavigator.setVisibilityChangedListener(new ActionProvider.VisibilityListener() {
            @Override
            public void onActionProviderVisibilityChanged(boolean isVisible) {
                if(isVisible)
                    topLayout.setVisibility(View.VISIBLE);
                else
                    topLayout.setVisibility(View.INVISIBLE);
            }
        });
    }

    private void hideSystemBars(){
        View decorView =  requireActivity().getWindow().getDecorView();
        WindowInsetsController windowInsetsController = decorView.getWindowInsetsController();
        if (windowInsetsController == null) {
            return;
        }
        // Configure the behavior of the hidden system bars
        windowInsetsController.setSystemBarsBehavior(
                WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        );
        getActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);

        //viewModel.getSystemBarsHidden().setValue(true);
    }

    private void showSystemBars(){
        View decorView =  requireActivity().getWindow().getDecorView();
        WindowInsetsController windowInsetsController = decorView.getWindowInsetsController();
        if (windowInsetsController == null) {
            return;
        }
        // Configure the behavior of the hidden system bars
        windowInsetsController.setSystemBarsBehavior(
                WindowInsetsController.BEHAVIOR_DEFAULT
        );
        windowInsetsController.show(WindowInsets.Type.systemBars());
        //viewModel.getSystemBarsHidden().setValue(false);
        requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);

    }

    @Override
    public FileViewerNavigator getNavigator() {
        return fileNavigator;
    }

    @Override
    public void onDetach() {
        super.onDetach();

        //((EnhancedMainMenuActivity)getActivity()).showStatusBar();
        //showSystemBars();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        showSystemBars();
        if(((SecureActivity)requireActivity()).getSupportActionBar() != null) {
            ((SecureActivity) requireActivity()).getSupportActionBar().show();
        }

    }

}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/ui/enhancedfileviewer/EnhancedFileViewFragment.kt
package com.quigglesproductions.secureimageviewer.ui.enhancedfileviewer

import android.content.Context
import android.os.Bundle
import android.view.ContextThemeWrapper
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.PopupMenu
import android.widget.TextView
import android.widget.ToggleButton
import androidx.annotation.OptIn
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.ActionProvider
import androidx.fragment.app.activityViewModels
import androidx.hilt.navigation.fragment.hiltNavGraphViewModels
import androidx.lifecycle.lifecycleScope
import androidx.media3.common.util.UnstableApi
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.quigglesproductions.secureimageviewer.R
import com.quigglesproductions.secureimageviewer.managers.VideoPlaybackManager
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.relations.RoomUnifiedEmbeddedFile
import com.quigglesproductions.secureimageviewer.ui.IFileViewer
import com.quigglesproductions.secureimageviewer.ui.SecureFragment
import com.quigglesproductions.secureimageviewer.ui.compoundcontrols.FileViewerNavigator
import com.quigglesproductions.secureimageviewer.ui.enhancedfolderlist.FolderListType
import com.quigglesproductions.secureimageviewer.ui.enhancedfolderviewer.FolderViewerViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class EnhancedFileViewFragment : SecureFragment(), IFileViewer {
    private var startPos = 0
    private lateinit var viewPager: ViewPager2
    lateinit var topLayout: ConstraintLayout
    private lateinit var backButton: ImageButton
    private lateinit var optionsButton: ImageButton
    private lateinit var favouriteButton: ToggleButton
    lateinit var fileName: TextView
    @UnstableApi
    lateinit var fileNavigator: FileViewerNavigator
    private var currentPagerSlopMultiplier = 0
    private var hasStartPosition: Boolean = false
    private val viewModel: EnhancedFileViewerViewModel by activityViewModels<EnhancedFileViewerViewModel>()
    private val folderViewModel by hiltNavGraphViewModels<FolderViewerViewModel>(R.id.main_navigation)
    private lateinit var collectionAdapter: EnhancedFileCollectionAdapterKt<IDisplayFile>

    @Inject
    lateinit var videoPlaybackManager: VideoPlaybackManager

    private val args: EnhancedFileViewFragmentArgs by navArgs()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val wrapper = ContextThemeWrapper(requireContext(), R.style.FileViewerTheme)
        val themedInflater = inflater.cloneInContext(wrapper)

        return themedInflater.inflate(R.layout.fragment_file_pager, container, false)
    }

    @OptIn(UnstableApi::class)
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        viewPager = view.findViewById(R.id.fragment_view_pager)
        if (!hasStartPosition)
            startPos = args.startPosition
        setupNavigationControls(view)
        collectionAdapter = EnhancedFileCollectionAdapterKt<IDisplayFile>(this,videoPlaybackManager,viewModel)
        collectionAdapter.setFileNavigator(fileNavigator)
        collectionAdapter.addFiles(folderViewModel.files.value)
        viewPager.adapter = collectionAdapter
        super.onViewCreated(view, savedInstanceState)
        //hideSystemBars()
        fileNavigator.setFileTotal(collectionAdapter.itemCount)
        viewPager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                //selectedFile = collectionAdapter!!.getItem(position)
                val selectedFile = folderViewModel.files.value?.get(position)
                fileName.text = selectedFile?.name
                folderViewModel.selectedFile.value = selectedFile
                //fileName!!.text = selectedFile!!.getName()
                fileNavigator.setFilePosition(position + 1)
                topLayout.invalidate()
                favouriteButton.isChecked = selectedFile!!.file.isFavourite
            }
        })
        viewPager.setCurrentItem(collectionAdapter.getPosition(folderViewModel.selectedFile.value), false)
        fileName.text = folderViewModel.selectedFile.value!!.getName()
        viewPager.setNestedScrollingEnabled(true)
        collectionAdapter.setFileZoomLevelCallback(EnhancedFileCollectionAdapterKt.ZoomLevelChangeCallback { isZoomed ->
            if (isZoomed) setViewPagerSlop(11) else setViewPagerSlop(1)
            return@ZoomLevelChangeCallback
        })
    }

    private fun setViewPagerSlop(multiplier: Int) {
        val initialValue = 42
        try {
            if (currentPagerSlopMultiplier != multiplier) {
                val recyclerViewField = viewPager.javaClass.getDeclaredField("mRecyclerView")
                recyclerViewField.isAccessible = true
                val recyclerView = recyclerViewField[viewPager] as RecyclerView
                val field = RecyclerView::class.java.getDeclaredField("mTouchSlop")
                field.isAccessible = true
                field[recyclerView] = initialValue * multiplier
            }
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } finally {
            currentPagerSlopMultiplier = multiplier
        }
    }

    @OptIn(UnstableApi::class)
    private fun setupNavigationControls(view: View) {
        topLayout = view.findViewById(R.id.topLinearLayout)
        backButton = topLayout.findViewById(R.id.backButton)
        optionsButton = topLayout.findViewById(R.id.optionsButton)
        favouriteButton = topLayout.findViewById(R.id.favouriteButton)
        fileName = topLayout.findViewById(R.id.file_title)
        fileNavigator = view.findViewById(R.id.fileviewer_navigator)
        backButton.setOnClickListener { requireActivity().onBackPressed() }
        fileNavigator.setNextButtonOnClickListener {
            viewPager.setCurrentItem(
                viewPager.currentItem + 1,
                true
            )
        }
        optionsButton.setOnClickListener(View.OnClickListener {
            val popup = PopupMenu(requireContext(),optionsButton,Gravity.END)
            popup.menuInflater.inflate(R.menu.menu_file_view,popup.menu)
            popup.setOnMenuItemClickListener {
                when(it.itemId){
                    R.id.action_info_view -> showFileInfoDialog()
                }
                return@setOnMenuItemClickListener true
            }
            popup.show()
        })
        favouriteButton.setOnCheckedChangeListener { buttonView, isChecked ->
            if(isChecked != folderViewModel.selectedFile.value!!.file.isFavourite) {
                folderViewModel.selectedFile.value!!.file.isFavourite = isChecked
                when (folderViewModel.folderListType.value) {
                    FolderListType.ONLINE -> {}
                    FolderListType.DOWNLOADED -> {
                        lifecycleScope.launch {
                            downloadFileDatabase.fileDao().update(folderViewModel.selectedFile.value!!.file)
                        }
                    }

                    else -> {}
                }
            }
        }
        fileNavigator.setPreviousButtonOnClickListener(View.OnClickListener {
            viewPager.setCurrentItem(
                viewPager.currentItem - 1,
                true
            )
        })
        fileNavigator.setVisibilityChangedListener(ActionProvider.VisibilityListener { isVisible ->
            if (isVisible) topLayout.visibility = View.VISIBLE else topLayout.visibility = View.INVISIBLE
        })
    }

    private fun showFileInfoDialog(){
        val bottomSheetDialog = BottomSheetDialog(requireContext())
        bottomSheetDialog.setContentView(R.layout.bottomdialog_fileinfo)
        bottomSheetDialog.behavior.state = BottomSheetBehavior.STATE_EXPANDED
        val itemNameText = bottomSheetDialog.findViewById<TextView>(R.id.item_name)
        val folderNameText = bottomSheetDialog.findViewById<TextView>(R.id.folder_name)
        val artistNameText = bottomSheetDialog.findViewById<TextView>(R.id.artist_name)
        val catagoriesText = bottomSheetDialog.findViewById<TextView>(R.id.catagories)
        val subjectsText = bottomSheetDialog.findViewById<TextView>(R.id.subjects)
        folderViewModel.selectedFile.value!!.dataSource.getFileMetadata(requiresRequestManager()) { metadata, exception -> //selectedFile.metadata = metadata;
            itemNameText!!.text = folderViewModel.selectedFile.value!!.name
            folderNameText!!.text = if(folderViewModel.selectedFile.value!!.file.cachedFolderName != null) folderViewModel.selectedFile.value!!.file.cachedFolderName else folderViewModel.folder.value?.name
            artistNameText!!.text = folderViewModel.selectedFile.value!!.artistName
            catagoriesText!!.text = folderViewModel.selectedFile.value!!.catagoryListString
            subjectsText!!.text = folderViewModel.selectedFile.value!!.subjectListString
            bottomSheetDialog.create()
            bottomSheetDialog.show()
        }
    }

    @OptIn(UnstableApi::class)
    override fun getNavigator(): FileViewerNavigator {
        return fileNavigator
    }
    override fun onDestroyView() {
        super.onDestroyView()

    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
    }

    override fun onDetach() {
        super.onDetach()
        requiresSecureActivity().showSystemUI()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        requiresSecureActivity().hideSystemUI()
    }

    fun updateFileFavourite(file: RoomUnifiedEmbeddedFile, state: Boolean) {
        file.file.isFavourite = state
        lifecycleScope.launch {
            downloadFileDatabase.fileDao().update(file = file.file)
        }
    }

    val zoomCallback: EnhancedFileCollectionAdapterKt.ZoomLevelChangeCallback
        get() = collectionAdapter.zoomLevelCallback

}
